candid_diff/src/candiff.rs

```diff
@@ -0,0 +1,302 @@
+        } => {
+            match format {
+                CliFormat::Text => {
+                    let ty = match input_type {
+                        None => {
+                            warn!("No input type provided; All numbers remain textual.");
+                            None
+                        }
+                        Some(t) => match t.parse::<IDLType>() {
+                            Ok(ty) => {
+                                trace!("input_type = {:?}", ty);
+                                Some(ty)
+                            }
+                            Err(e) => {
+                                error!("{}", e);
+                                None
+                            }
+                        },
+                    };
+                    match Value::from_str(&input) {
+                        Ok(v) => {
+                            // check the type, and then annotate the value with that type, possibly transforming it
+                            let v = match ty {
+                                None => v,
+                                Some(ty) => {
+                                    // to do -- check a program file and get the type env
+                                    let te = &mut TypeEnv::new();
+                                    trace!("check_type ...");
+                                    match check_type(&Env { te, pre: false }, &ty) {
+                                        Err(e) => {
+                                            error!("{}", e);
+                                            type_consistency_holds = false;
+                                            v
+                                        }
+                                        Ok(ty) => {
+                                            trace!("annotate_type ...");
+                                            match v.annotate_type(&TypeEnv::new(), &ty) {
+                                                Ok(v) => v,
+                                                Err(e) => {
+                                                    type_consistency_holds = false;
+                                                    error!("{}", e);
+                                                    v
+                                                }
+                                            }
+                                        }
+                                    }
+                                }
+                            };
+                            trace!("{:?}", v);
+                            if debug_output {
+                                println!("{:?}", v)
+                            } else {
+                                println!("{}", v)
+                            }
+                        }
+                        Err(e) => error!("{}", e),
+                    }
+                }
+                CliFormat::Raw => {
+                    // how is it encoded as a string?
```

Comment:
If we are following the format of `dfx --output raw`, `hex::encode(raw_string)` will be fine.

Reply:
Okay, good to know.

This PR is getting gigantic, so I may leave that for another one.